File: archiso/airootfs/usr/share/churros/preferences/pages/wallpaper.py
import os

import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk, Gdk, GLib, Gio

from widgets.page import Page
from widgets.group import Group
from widgets.row import Row

from services.wallpaper import WallpaperService


class WallpaperPage(Page):

    def __init__(self, navigator):

        super().__init__(
            navigator,
            "Fondos",
            "Selecciona un fondo de pantalla",
            parent_page="appearance"
        )

        self.navigator = navigator

        #
        # Botón "Importar..."
        #

        actions_group = Group("Importar fondo")
        actions_group.add(
            Row(
                title="Importar desde archivos...",
                subtitle="Abre el selector de archivos de GTK",
                icon="wallpaper.svg",
                callback=lambda *_: self.import_from_files()
            )
        )
        self.add(actions_group)

        thunar_group = Group("Abrir carpeta")
        thunar_group.add(
            Row(
                title="Abrir ~/Imagenes con Thunar",
                subtitle="Arrastra fondos a ~/.local/share/churros/wallpapers",
                icon="wallpaper.svg",
                callback=lambda *_: self._open_pictures_folder()
            )
        )
        self.add(thunar_group)

        #
        # Fondo actual + grid
        #

        current = WallpaperService.current()
        wallpapers = WallpaperService.available()

        if not wallpapers:

            group = Group("Fondos disponibles")
            group.add(
                Row(
                    title="No se encontraron fondos",
                    subtitle="Importa una imagen o añádela a ~/.local/share/churros/wallpapers",
                    icon="wallpaper.svg"
                )
            )
            self.add(group)
            return

        #
        # Miniatura del fondo actual
        #

        current_group = Group("Fondo actual")

        if current and os.path.exists(current):

            try:

                texture = Gdk.Texture.new_from_filename(current)
                preview = Gtk.Image.new_from_paintable(texture)
                preview.set_pixel_size(160)
                preview.add_css_class("wallpaper-preview")

                current_group.add(
                    Row(
                        title=os.path.splitext(
                            os.path.basename(current)
                        )[0],
                        subtitle="Seleccionado",
                        icon="wallpaper.svg"
                    )
                )

                self.add(current_group)

            except Exception:

                self.add(current_group)

        #
        # Grid de fondos
        #

        grid_group = Group("Fondos disponibles")

        flow = Gtk.FlowBox()
        flow.set_selection_mode(Gtk.SelectionMode.NONE)
        flow.set_max_children_per_line(4)
        flow.set_min_children_per_line(2)
        flow.set_row_spacing(12)
        flow.set_column_spacing(12)
        flow.set_halign(Gtk.Align.FILL)

        for wallpaper in wallpapers:

            thumb = self._build_thumbnail(
                wallpaper,
                current
            )

            flow.insert(thumb, -1)

        grid_group.add(flow)

        self.add(grid_group)

    def import_from_files(self):

        win = self.get_root()

        try:
            dialog = Gtk.FileDialog()
            dialog.set_title("Importar imagen de fondo")

            filter_any = Gtk.FileFilter()
            filter_any.set_name("Imagenes")
            filter_any.add_mime_type("image/jpeg")
            filter_any.add_mime_type("image/png")
            filter_any.add_mime_type("image/webp")
            filter_any.add_mime_type("image/gif")
            filter_any.add_pattern("*.jpg")
            filter_any.add_pattern("*.jpeg")
            filter_any.add_pattern("*.png")
            filter_any.add_pattern("*.webp")
            filter_any.add_pattern("*.gif")

            filters = Gio.ListStore.new(Gtk.FileFilter)
            filters.append(filter_any)
            dialog.set_filters(filters)
            dialog.set_default_filter(filter_any)

            try:
                dialog.set_initial_folder(
                    Gio.File.new_for_path(os.path.expanduser("~"))
                )
            except Exception:
                pass

            def on_result(source, result, _user_data=None):
                try:
                    file = dialog.open_finish(result)
                except GLib.Error:
                    return

                if file is None:
                    return

                src = file.get_path() if hasattr(file, "get_path") else None

                if not src or not os.path.isfile(src):
                    return

                self._apply_wallpaper(src)

            dialog.open(win, None, on_result)

        except Exception:
            self._open_pictures_folder()

    def _open_pictures_folder(self):
        import subprocess

        pics_dir = os.path.expanduser("~/Imagenes")
        if not os.path.isdir(pics_dir):
            pics_dir = os.path.expanduser("~")

        try:
            subprocess.Popen(
                ["thunar", pics_dir],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                start_new_session=True,
            )
        except Exception:
            pass

    def _apply_wallpaper(self, src):

        dest = WallpaperService.import_image(src)

        if dest is None:
            print("[wallpaper] no se pudo importar", src, flush=True)
            self._show_error("No se pudo importar el fondo", src)
            return

        success = WallpaperService.set(dest)

        print("[wallpaper] import+set retorno:", success, "dest:", dest, flush=True)

        if not success:
            self._show_error(
                "No se pudo aplicar el fondo",
                "Revisa /tmp/churros-settings.log y /tmp/swaybg.log"
            )

        self._rebuild_grid()

    def _show_error(self, message, detail):

        try:

            win = self.get_root()

            dialog = Gtk.AlertDialog()

            dialog.set_modal(True)

            dialog.set_message(message)

            dialog.set_detail(detail)

            if win is not None:
                dialog.show(win)
            else:
                dialog.show()

        except Exception:
            pass

    def _rebuild_grid(self):
        """Recarga todo el contenido de la pagina para mostrar la nueva imagen."""

        content = self.content
        child = content.get_first_child()
        while child is not None:
            nxt = child.get_next_sibling()
            content.remove(child)
            child = nxt

        # Re-construir (re-ejecutar la parte grafica del __init__)
        self._build_after_import()

    def _build_after_import(self):
        """Reconstruye los grupos tras importar un fondo nuevo."""

        actions_group = Group("Importar fondo")
        actions_group.add(
            Row(
                title="Importar desde archivos...",
                subtitle="Elige una imagen de tu disco duro",
                icon="wallpaper.svg",
                callback=lambda *_: self.import_from_files()
            )
        )
        self.add(actions_group)

        current = WallpaperService.current()
        wallpapers = WallpaperService.available()

        if not wallpapers:

            group = Group("Fondos disponibles")
            group.add(
                Row(
                    title="No se encontraron fondos",
                    subtitle="Importa una imagen",
                    icon="wallpaper.svg"
                )
            )
            self.add(group)
            return

        current_group = Group("Fondo actual")

        if current and os.path.exists(current):

            try:

                texture = Gdk.Texture.new_from_filename(current)
                preview = Gtk.Image.new_from_paintable(texture)
                preview.set_pixel_size(160)
                preview.add_css_class("wallpaper-preview")

                current_group.add(
                    Row(
                        title=os.path.splitext(
                            os.path.basename(current)
                        )[0],
                        subtitle="Seleccionado",
                        icon="wallpaper.svg"
                    )
                )

                self.add(current_group)

            except Exception:
                self.add(current_group)

        grid_group = Group("Fondos disponibles")
        flow = Gtk.FlowBox()
        flow.set_selection_mode(Gtk.SelectionMode.NONE)
        flow.set_max_children_per_line(4)
        flow.set_min_children_per_line(2)
        flow.set_row_spacing(12)
        flow.set_column_spacing(12)
        flow.set_halign(Gtk.Align.FILL)

        for wallpaper in wallpapers:

            thumb = self._build_thumbnail(wallpaper, current)
            flow.insert(thumb, -1)

        grid_group.add(flow)

        self.add(grid_group)

    def _build_thumbnail(self, wallpaper, current):

        box = Gtk.Box(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=6
        )

        name = os.path.splitext(
            os.path.basename(wallpaper)
        )[0]

        is_current = (wallpaper == current)

        try:

            texture = Gdk.Texture.new_from_filename(wallpaper)
            image = Gtk.Image.new_from_paintable(texture)
            image.set_pixel_size(120)
            image.add_css_class("wallpaper-thumb")

            if is_current:
                image.add_css_class("wallpaper-selected")

        except Exception:

            image = Gtk.Image.new_from_icon_name("image-missing")
            image.set_pixel_size(120)

        button = Gtk.Button()
        button.set_child(image)
        button.add_css_class("wallpaper-button")
        button.set_has_frame(False)
        button.set_tooltip_text(name)
        button.connect(
            "clicked",
            lambda _, w=wallpaper: self.select(w)
        )

        label = Gtk.Label(label=name)
        label.add_css_class("wallpaper-name")
        label.set_max_width_chars(18)
        label.set_ellipsize(0)  # PANGO_ELLIPSIZE_END
        label.set_tooltip_text(name)

        box.append(button)
        box.append(label)

        return box

    def select(self, wallpaper):

        print("[wallpaper-page] seleccion:", wallpaper, flush=True)

        success = WallpaperService.set(wallpaper)

        print("[wallpaper-page] set retorno:", success, flush=True)

        if not success:

            try:

                win = self.get_root()

                dialog = Gtk.AlertDialog()

                dialog.set_modal(True)

                dialog.set_message("No se pudo aplicar el fondo")

                dialog.set_detail(

                    "Revisa /tmp/churros-settings.log, "

                    "/tmp/swaybg.log"

                )

                if win is not None:

                    dialog.show(win)

                else:

                    dialog.show()

            except Exception:

                pass

        GLib.idle_add(
            lambda: self.navigator.show_page("appearance")
        )

File: rust/preferences/src/pages/wallpaper.rs
// ==========================================
// WallpaperPage — fondos de pantalla (equivalente a pages/wallpaper.py)
// ==========================================

use std::path::Path;
use std::process::Command;

use gtk::prelude::*;

use crate::services::wallpaper::WallpaperService;
use crate::widgets::group::Group;
use crate::widgets::page::Page;
use crate::widgets::row::Row;

pub fn build(navigator: gtk::Stack) -> Page {
    let page = Page::new(
        Some(navigator.clone()),
        "Fondos",
        Some("Selecciona un fondo de pantalla"),
        Some("appearance".to_string()),
    );

    // Contenido reutilizable para reconstruir la página tras importar (self.content)
    let content: gtk::Box = page.content.clone();

    // ===== Botón "Importar..." =====
    build_actions(&page, &navigator, &content);

    // ===== Fondo actual + grid =====
    let current = WallpaperService::current();
    let wallpapers = WallpaperService::list();

    if wallpapers.is_empty() {
        let mut group = Group::new("Fondos disponibles");
        group.add(&Row::new(
            "No se encontraron fondos",
            Some("Importa una imagen o añádela a ~/.local/share/churros/wallpapers"),
            Some("wallpaper.svg"),
            None,
            None,
            None,
        ));
        page.add(group.widget());
        return page;
    }

    // ===== Miniatura del fondo actual =====
    if !current.is_empty() && Path::new(&current).is_file() {
        let mut current_group = Group::new("Fondo actual");

        // El Python carga Gdk.Texture.new_from_filename + crea un Gtk.Image con la
        // clase "wallpaper-preview"... pero NUNCA lo añade al grupo: solo añade la Row.
        // Código muerto del original; se omite la imagen (try/except silencioso).
        let name = Path::new(&current)
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("")
            .to_string();

        current_group.add(&Row::new(
            &name,
            Some("Seleccionado"),
            Some("wallpaper.svg"),
            None,
            None,
            None,
        ));

        page.add(current_group.widget());
    }

    // ===== Grid de fondos =====
    build_grid(&content, &current, &wallpapers, &navigator);

    page
}

/// Grupos "Importar fondo" + "Abrir carpeta" (equivalente al inicio de __init__)
fn build_actions(page: &Page, navigator: &gtk::Stack, content: &gtk::Box) {
    let mut actions_group = Group::new("Importar fondo");

    let nav = navigator.clone();
    let content_cb = content.clone();
    actions_group.add(&Row::new(
        "Importar desde archivos...",
        Some("Abre el selector de archivos de GTK"),
        Some("wallpaper.svg"),
        None,
        None,
        Some(Box::new(move |_| {
            import_from_files(&nav, &content_cb);
        })),
    ));

    page.add(actions_group.widget());

    let mut thunar_group = Group::new("Abrir carpeta");

    let nav = navigator.clone();
    thunar_group.add(&Row::new(
        "Abrir carpeta de fondos",
        Some("Arrastra fondos a ~/.local/share/churros/wallpapers"),
        Some("wallpaper.svg"),
        None,
        None,
        Some(Box::new(move |_| {
            open_pictures_folder();
        })),
    ));

    page.add(thunar_group.widget());
}

/// Grid de fondos con FlowBox (equivalente a la parte final de __init__
/// y de _build_after_import)
fn build_grid(content: &gtk::Box, current: &str, wallpapers: &[std::path::PathBuf], navigator: &gtk::Stack) {
    let mut grid_group = Group::new("Fondos disponibles");

    let flow = gtk::FlowBox::new();
    flow.set_selection_mode(gtk::SelectionMode::None);
    flow.set_max_children_per_line(4);
    flow.set_min_children_per_line(2);
    flow.set_row_spacing(12);
    flow.set_column_spacing(12);
    flow.set_halign(gtk::Align::Fill);

    for wallpaper in wallpapers {
        let thumb = build_thumbnail(wallpaper, current, navigator);
        flow.insert(&thumb, -1);
    }

    // Gtk::FlowBox no implementa AsWidget; se inserta directo en la tarjeta del grupo
    grid_group.card.append(&flow);

    content.append(grid_group.widget());
}

/// Miniatura: botón con imagen + nombre (equivalente a _build_thumbnail)
fn build_thumbnail(wallpaper: &Path, current: &str, navigator: &gtk::Stack) -> gtk::Box {
    let box_ = gtk::Box::new(gtk::Orientation::Vertical, 6);

    let name = wallpaper
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("")
        .to_string();

    let is_current = wallpaper.to_string_lossy() == current;

    let image = match gtk::gdk::Texture::from_filename(wallpaper) {
        Ok(texture) => {
            let image = gtk::Image::from_paintable(Some(&texture));
            image.set_pixel_size(120);
            image.add_css_class("wallpaper-thumb");
            if is_current {
                image.add_css_class("wallpaper-selected");
            }
            image
        }
        Err(_) => {
            let image = gtk::Image::from_icon_name("image-missing");
            image.set_pixel_size(120);
            image
        }
    };

    let button = gtk::Button::new();
    button.set_child(Some(&image));
    button.add_css_class("wallpaper-button");
    button.set_has_frame(false);
    button.set_tooltip_text(Some(&name));

    let target = wallpaper.to_string_lossy().to_string();
    let nav = navigator.clone();
    button.connect_clicked(move |_| {
        select(&target, &nav);
    });

    let label = gtk::Label::new(Some(&name));
    label.add_css_class("wallpaper-name");
    label.set_max_width_chars(18);
    // Python: set_ellipsize(0) == PANGO_ELLIPSIZE_NONE (el comentario del original
    // dice END, pero 0 es NONE) — se porta el valor real: NONE.
    label.set_ellipsize(gtk::pango::EllipsizeMode::None);
    label.set_tooltip_text(Some(&name));

    box_.append(&button);
    box_.append(&label);

    box_
}

/// Ventana raíz de la página (equivalente a self.get_root() del Python)
fn root_window(navigator: &gtk::Stack) -> Option<gtk::Window> {
    navigator.root().and_then(|r| r.downcast::<gtk::Window>().ok())
}

/// Equivalente a WallpaperPage.import_from_files: FileDialog asíncrono
fn import_from_files(navigator: &gtk::Stack, content: &gtk::Box) {
    let Some(win) = root_window(navigator) else {
        show_error(navigator, "No se pudo abrir el selector", "La pagina no tiene ventana root.");
        return;
    };

    let dialog = gtk::FileDialog::new();
    dialog.set_title("Importar imagen de fondo");
    dialog.set_modal(true);

    let filter_any = gtk::FileFilter::new();
    filter_any.set_name(Some("Imagenes"));
    filter_any.add_mime_type("image/jpeg");
    filter_any.add_mime_type("image/png");
    filter_any.add_mime_type("image/webp");
    filter_any.add_mime_type("image/gif");
    filter_any.add_pattern("*.jpg");
    filter_any.add_pattern("*.jpeg");
    filter_any.add_pattern("*.png");
    filter_any.add_pattern("*.webp");
    filter_any.add_pattern("*.gif");

    let filters = gio::ListStore::new::<gtk::FileFilter>();
    filters.append(&filter_any);
    dialog.set_filters(Some(&filters));
    dialog.set_default_filter(Some(&filter_any));

    // set_initial_folder(home) — el Python lo envuelve en try/except silencioso
    let home = std::env::var("HOME").unwrap_or_else(|_| "/root".to_string());
    let home_file = gio::File::for_path(&home);
    dialog.set_initial_folder(Some(&home_file));

    let nav = navigator.clone();
    let content_cb = content.clone();
    dialog.open(Some(&win), None::<&gio::Cancellable>, move |result| {
        // Equivalente a on_result: open_finish
        let file = match result {
            Ok(f) => f,
            Err(e) => {
                println!("[wallpaper] FileDialog error: {e}");
                return;
            }
        };

        let Some(src) = file.path() else {
            println!("[wallpaper] ruta invalida: None");
            return;
        };
        let src = src.to_string_lossy().to_string();

        if !Path::new(&src).is_file() {
            println!("[wallpaper] ruta invalida: {src}");
            return;
        }

        apply_wallpaper(&src, &nav, &content_cb);
    });
}

/// Equivalente a _open_pictures_folder: crea ~/.local/share/churros/wallpapers
/// y la abre con thunar (fallback xdg-open), todo silencioso.
fn open_pictures_folder() {
    let wp_dir = WallpaperService::user_dir();
    let _ = std::fs::create_dir_all(&wp_dir);

    let dir_str = wp_dir.to_string_lossy().to_string();

    if Command::new("thunar")
        .arg(&dir_str)
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .spawn()
        .is_err()
    {
        let _ = Command::new("xdg-open")
            .arg(&dir_str)
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .spawn();
    }
}

/// Equivalente a _apply_wallpaper: importar + aplicar + reconstruir el grid
fn apply_wallpaper(src: &str, navigator: &gtk::Stack, content: &gtk::Box) {
    let Some(dest) = WallpaperService::import_image(src) else {
        println!("[wallpaper] no se pudo importar {src}");
        show_error(navigator, "No se pudo importar el fondo", src);
        return;
    };

    rebuild_grid(content, navigator);

    let dest_clone = dest.clone();
    std::thread::spawn(move || {
        let success = WallpaperService::set(&dest_clone);
        println!("[wallpaper] import+set retorno: {success} dest: {dest_clone}");
    });
}

/// Equivalente a _show_error: Gtk::AlertDialog modal. Si no hay ventana raíz no
/// se muestra nada (en el Python dialog.show() sin ventana falla y el try/except
/// lo traga).
fn show_error(navigator: &gtk::Stack, message: &str, detail: &str) {
    let dialog = gtk::AlertDialog::builder()
        .modal(true)
        .message(message)
        .detail(detail)
        .build();

    if let Some(win) = root_window(navigator) {
        dialog.show(Some(&win));
    }
}

/// Equivalente a _rebuild_grid: vacía el contenido de la página y reconstruye
/// los grupos (re-ejecuta la parte gráfica del __init__).
fn rebuild_grid(content: &gtk::Box, navigator: &gtk::Stack) {
    // Vaciar el contenido (while child is not None: remove)
    let mut child = content.first_child();
    while let Some(c) = child {
        let nxt = c.next_sibling();
        content.remove(&c);
        child = nxt;
    }

    build_after_import(content, navigator);
}

/// Equivalente a _build_after_import: reconstruye los grupos tras importar.
/// NOTA: el Python solo re-añade el grupo "Importar fondo" (el de "Abrir carpeta"
/// desaparece tras una importación — bug del original, se porta tal cual).
fn build_after_import(content: &gtk::Box, navigator: &gtk::Stack) {
    let mut actions_group = Group::new("Importar fondo");

    let nav = navigator.clone();
    let content_cb = content.clone();
    actions_group.add(&Row::new(
        "Importar desde archivos...",
        Some("Elige una imagen de tu disco duro"),
        Some("wallpaper.svg"),
        None,
        None,
        Some(Box::new(move |_| {
            import_from_files(&nav, &content_cb);
        })),
    ));

    content.append(actions_group.widget());

    let current = WallpaperService::current();
    let wallpapers = WallpaperService::list();

    if wallpapers.is_empty() {
        let mut group = Group::new("Fondos disponibles");
        group.add(&Row::new(
            "No se encontraron fondos",
            Some("Importa una imagen"),
            Some("wallpaper.svg"),
            None,
            None,
            None,
        ));
        content.append(group.widget());
        return;
    }

    // Fondo actual
    if !current.is_empty() && Path::new(&current).is_file() {
        let mut current_group = Group::new("Fondo actual");

        let name = Path::new(&current)
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("")
            .to_string();

        current_group.add(&Row::new(
            &name,
            Some("Seleccionado"),
            Some("wallpaper.svg"),
            None,
            None,
            None,
        ));

        content.append(current_group.widget());
    }

    // Grid
    build_grid(content, &current, &wallpapers, navigator);
}

/// Equivalente a WallpaperPage.select: aplicar fondo + volver a apariencia.
/// Se ejecuta en un hilo secundario para no bloquear el loop de eventos Wayland de GTK4.
fn select(wallpaper: &str, navigator: &gtk::Stack) {
    println!("[wallpaper-page] seleccion: {wallpaper}");

    let nav = navigator.clone();
    let wp = wallpaper.to_string();

    // Navegación inmediata a Apariencia para respuesta visual instantánea
    glib::idle_add_local_once(move || {
        nav.set_visible_child_name("appearance");
    });

    // Aplicar wallpaper y colores pywal en segundo plano de forma no bloqueante
    std::thread::spawn(move || {
        let success = WallpaperService::set(&wp);
        println!("[wallpaper-page] set retorno: {success}");
    });
}
